gcc/rust/backend/rust-constexpr.cc

```diff
@@ -2953,14 +2953,14 @@ eval_store_expression (const constexpr_ctx *ctx, tree t, bool lval,
       TREE_SIDE_EFFECTS (*valp) = TREE_SIDE_EFFECTS (init);
       CONSTRUCTOR_NO_CLEARING (*valp) = CONSTRUCTOR_NO_CLEARING (init);
     }
-  else if (TREE_CODE (init) == CONSTRUCTOR
-	   && !same_type_ignoring_top_level_qualifiers_p (TREE_TYPE (init),
-							  type))
-    {
-      /* See above on initialization of empty bases.  */
-      gcc_assert (is_empty_class (TREE_TYPE (init)) && !lval);
-      return init;
-    }
+  // else if (TREE_CODE (init) == CONSTRUCTOR
+  //          && !same_type_ignoring_top_level_qualifiers_p (TREE_TYPE (init),
+  //       						  type))
+  //   {
+  //     /* See above on initialization of empty bases.  */
+  //     // gcc_assert (is_empty_class (TREE_TYPE (init)) && !lval);
+  //     return init;
+  //   }
```

Comment:
Dead code? Or will that be useful later on?

Reply:
I might want this later on, not 100% sure. Basically we don't really set TYPE_CANONICAL and the qualifiers differ here. So for example:

```
let a:[i32,5] = {1,2,3,4,5}
```

We end up with two types in the type system one for the lhs and one for the rhs they are equal (both array types with the same element and capacity). We do have the type hasher which means we avoid needing to use TYPE_CANONICAL because we try really hard to just use the same type as much as possible avoiding duplicates. Otherwise, GCC (I think this might be the fault of the GO abstractions) generate a whole bunch of VIEW_CONVERT_EXPR's to convert between the equivalent structures.

But one has the qualifiers of const and one is not marked const (since the lhs is not mutable) so we end up hitting this !same_type_ignoring_top_level_qualifiers_p and drop into this where yeah this is not an empty class. I am loathed to remove this code only because it makes it easier to debug and see what code we changed in our version of constexpr compared to the C++ one.

We could get around this by avoiding setting the const qualifiers on immutable let bindings but I think its good to give GCC more information for the optimizers to hopefully work well.